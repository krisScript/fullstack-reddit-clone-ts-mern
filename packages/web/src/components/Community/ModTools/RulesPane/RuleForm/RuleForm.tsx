import React, { FC, Dispatch, SetStateAction, SyntheticEvent } from 'react';
import { FormComponentProps } from 'antd/lib/form/Form';
import ValidationError from '@rddt/common/types/ValidationError';
import Rule from '@rddt/common/types/Rule';
import RuleFormValues from 'types/RuleFormValues';
import { Form, Icon, Input, Button, Radio } from 'antd';
const { TextArea } = Input;
const { Group } = Radio;
interface WrappedRuleFormProps extends FormComponentProps {
  cancelHandler: () => void;
  addRuleHandler: (values: RuleFormValues) => Promise<void>;
  editRuleHandler: (values: RuleFormValues) => Promise<void>;
  type: 'Add' | 'Edit';
  editRule: Rule | undefined;
}
interface RuleFormProps extends WrappedRuleFormProps {
  setConfirmLoading: Dispatch<SetStateAction<boolean>>;
}
const { Item } = Form;
const RuleForm: FC<RuleFormProps> = ({
  form,
  setConfirmLoading,
  cancelHandler,
  addRuleHandler,
  editRuleHandler,
  type,
  editRule,
}) => {
  const resetHandler = (e: SyntheticEvent) => {
    e.preventDefault();
    form.resetFields();
  };
  const submitHandler = (e: SyntheticEvent) => {
    e.preventDefault();
    form.validateFields(async (err, values: any) => {
      try {
        setConfirmLoading(true);
        if (type === 'Add') {
          addRuleHandler(values);
        } else if (type === 'Edit' && editRule) {
          editRuleHandler(values);
        }
        setConfirmLoading(false);
        form.resetFields();
        cancelHandler();
      } catch (err) {
        setConfirmLoading(false);
        if (err.response.data.data) {
          const { data } = err.response.data;
          data.map((validationErr: ValidationError) => {
            form.setFields({
              [validationErr.param]: {
                value: validationErr.value,
                errors: [new Error(validationErr.msg)],
              },
            });
          });
        }
      }
    });
  };
  const { getFieldDecorator } = form;
  return (
    <Form onSubmit={submitHandler}>
      <Item>
        {getFieldDecorator('name', {
          initialValue: editRule && type === 'Edit' ? editRule.name : '',
          rules: [
            {
              required: true,
              message: 'Enter rule name!',
              min: 4,
              type: 'string',
            },
          ],
        })(<Input placeholder="Name" />)}
      </Item>
      <Item>
        {getFieldDecorator('description', {
          initialValue: editRule && type === 'Edit' ? editRule.description : '',
          rules: [
            {
              required: true,
              message: 'Enter rule description!',
              min: 4,
              type: 'string',
            },
          ],
        })(<TextArea rows={5} placeholder="Description" />)}
      </Item>
      <Item>
        {getFieldDecorator('scope', {
          initialValue: editRule && type === 'Edit' ? editRule.scope : '',
          rules: [
            {
              required: true,
              message: 'Enter rule description!',
              min: 4,
              type: 'string',
            },
          ],
        })(
          <Group>
            <Radio value={'Posts & comments'}>Posts and comments</Radio>
            <Radio value={'Posts only'}>Posts only</Radio>
            <Radio value={'Comments only'}>Comments only</Radio>
          </Group>,
        )}
      </Item>
      <Item>
        <Button type="primary" htmlType="submit" block>
          {type} Rule
        </Button>
        <Button htmlType="submit" onClick={resetHandler} block type="danger">
          Reset
        </Button>
      </Item>
    </Form>
  );
};

const WrappedRuleForm = Form.create<WrappedRuleFormProps>({ name: 'ruleForm' })(
  RuleForm,
);

export default WrappedRuleForm;
