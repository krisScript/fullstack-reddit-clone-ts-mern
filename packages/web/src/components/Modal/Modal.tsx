import React, { FC, cloneElement, useState } from 'react';
import { Modal as AntModal, Button } from 'antd';
interface ModalProps {
  title: string;
  cancelHandler: () => void;
}
const Modal: FC<ModalProps> = ({ title, cancelHandler, children }) => {
  const [confirmLoading, setConfirmLoading] = useState<boolean>(false);
  return (
    <AntModal
      title={title}
      visible={true}
      onCancel={cancelHandler}
      confirmLoading={confirmLoading}
      footer={[
        <Button
          key="back"
          loading={confirmLoading}
          onClick={cancelHandler}
          type="danger"
        >
          Cancel
        </Button>,
      ]}
    >
      {cloneElement(children as React.ReactElement<any>, {
        setConfirmLoading,
      })}
    </AntModal>
  );
};

export default Modal;
