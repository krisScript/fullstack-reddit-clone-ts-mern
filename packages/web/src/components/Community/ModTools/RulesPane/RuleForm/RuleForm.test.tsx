import React from 'react';
import Modal from 'components/Modal/Modal';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from '@testing-library/react';
import RuleForm from './RuleForm';
describe('<RuleForm />', () => {
  const nameValue = 'testValue';
  const descriptionValue = 'testDescription';
  const cancelHandler = jest.fn();
  const addRuleHandler = jest.fn();
  const editRuleHandler = jest.fn();
  const type = 'Add';
  const editRule = {
    name: 'Edit',
    description: 'Rule',
    scope: 'Posts Only',
    date: 'date',
    community: 'communityId',
    user: 'userId',
    _id: 'testId',
  };
  const { container, getByTestId, getByText, getByPlaceholderText } = render(
    <RuleForm
      cancelHandler={cancelHandler}
      editRule={editRule}
      addRuleHandler={addRuleHandler}
      editRuleHandler={editRuleHandler}
      type={type}
    />,
  );
  beforeEach(async () => {
    console.log(container);
    const nameInput = await waitForElement(() => getByPlaceholderText('Name'));
    fireEvent.change(nameInput, { target: { value: nameValue } });
    const descriptionInput = await waitForElement(() =>
      getByPlaceholderText('Description'),
    );
    fireEvent.change(descriptionInput, { target: { value: descriptionValue } });
    const scopeButton = getByText('Posts only');
    fireEvent.click(scopeButton);
    const submitButton = getByText('Add Rule');
    fireEvent.click(submitButton);
  });
  afterEach(cleanup);

  it('snapshot', () => {
    expect(container).toMatchSnapshot();
  });
  it('should call addRuleHandler once', async () => {
    expect(addRuleHandler).toHaveBeenCalledTimes(1);
  });
});
