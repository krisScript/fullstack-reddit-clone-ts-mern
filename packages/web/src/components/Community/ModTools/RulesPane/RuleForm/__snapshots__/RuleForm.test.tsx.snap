// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RuleForm /> snapshot 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-control-wrapper"
      >
        <div
          class="ant-form-item-control has-success"
        >
          <span
            class="ant-form-item-children"
          >
            <input
              class="ant-input"
              data-__field="[object Object]"
              data-__meta="[object Object]"
              id="ruleForm_name"
              placeholder="Name"
              type="text"
              value="testValue"
            />
          </span>
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-control-wrapper"
      >
        <div
          class="ant-form-item-control has-success"
        >
          <span
            class="ant-form-item-children"
          >
            <textarea
              class="ant-input"
              data-__field="[object Object]"
              data-__meta="[object Object]"
              id="ruleForm_description"
              placeholder="Description"
              rows="5"
            >
              testDescription
            </textarea>
          </span>
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-control-wrapper"
      >
        <div
          class="ant-form-item-control"
        >
          <span
            class="ant-form-item-children"
          >
            <div
              class="ant-radio-group ant-radio-group-outline"
              id="ruleForm_scope"
            >
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="Posts & comments"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  Posts and comments
                </span>
              </label>
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="Posts only"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  Posts only
                </span>
              </label>
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="Comments only"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  Comments only
                </span>
              </label>
            </div>
          </span>
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-control-wrapper"
      >
        <div
          class="ant-form-item-control"
        >
          <span
            class="ant-form-item-children"
          >
            <button
              ant-click-animating-without-extra-node="false"
              class="ant-btn ant-btn-primary ant-btn-block"
              type="submit"
            >
              <span>
                Add Rule
              </span>
            </button>
            <button
              class="ant-btn ant-btn-danger ant-btn-block"
              type="submit"
            >
              <span>
                Reset
              </span>
            </button>
          </span>
        </div>
      </div>
    </div>
  </form>
</div>
`;
